### PR TITLE
pallet-collective | port to frame v2 

### PR DESCRIPTION
**Pull request by shamb0**
# Plz kindly ignore this,

Changes moved to #8193

[#8193](https://github.com/paritytech/substrate/pull/8193)

---

# pallet-collective | port to frame v2 

related #7882

:warning: Breaking Change :warning:

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

```
storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.
```

So users of the collective pallet must be careful about the name they used in construct_runtime!. Hence the runtime-migration label, which might not be needed depending on the configuration of the collective pallet.

### CODE REVIEW

**Comment by shamb0** [comment]
Hello  @thiolliere ,

I hope Frame v2 migration is done, 
& ready for review.

Unit test & Benchmarking are passed.

Refactored the Unit test, similar to file structure used in pallet-balance.

Please kindly share your suggestions & comments.

**Comment by thiolliere** [comment]
It will need a polkadot companion or a migration because the pallet prefix used by storage is changing (in general)
also can you add this message to the top comment ?

## ⚠️ Breaking Change ⚠️ 

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the `ElectionsPhragmen` pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the `ElectionsPhragmen` pallet.

File: frame/collective/src/tests.rs [hunk 1]
```diff
@@ -0,0 +1,689 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Macro for creating the tests for the module.
+
+#![cfg(test)]
+
+#[macro_export]
+macro_rules! decl_tests {
```
Comment (thiolliere) [inline frame/collective/src/tests.rs:23]: we don't need to wrap test in a macro.
pallet-balance do that in order to run test on 2 different configuration, but here we have only one, thus it is not needed.

File: frame/collective/src/lib.rs [hunk 2]
```diff
@@ -955,778 +926,4 @@ impl<
 	fn successful_origin() -> O {
 		O::from(RawOrigin::Members(0u32, 0u32))
 	}
-}
-
-#[cfg(test)]
-mod tests {
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:961]: just to be sure, only trivial change were made in the test isn't it ?

File: frame/collective/src/benchmarking.rs [hunk 3]
```diff
@@ -35,18 +37,17 @@ use frame_system::Module as System;
...
-fn assert_last_event<T: Config<I>, I: Instance>(generic_event: <T as Config<I>>::Event) {
+fn assert_last_event<T: Config<I>, I: 'static>(generic_event: <T as Config<I>>::Event) {
 	let events = System::<T>::events();
 	let system_event: <T as frame_system::Config>::Event = generic_event.into();
 	// compare to the last event record
 	let EventRecord { event, .. } = &events[events.len() - 1];
 	assert_eq!(event, &system_event);
 }
 
-benchmarks_instance! {
```
Comment (thiolliere) [inline frame/collective/src/benchmarking.rs:49]: why did you change this to benchmark ?

NOTE: I think pallet-balance should make use of benchmarks_instance.

File: frame/collective/src/mock.rs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+	pub enum Test where
+		Block = Block,
+		NodeBlock = Block,
+		UncheckedExtrinsic = UncheckedExtrinsic
+	{
+		System: system::{Module, Call, Event<T>},
+		Collective: collective::<Instance1>::{Module, Call, Event<T>, Origin<T>, Config<T>},
+		CollectiveMajority: collective::<Instance2>::{Module, Call, Event<T>, Origin<T>, Config<T>},
+		DefaultCollective: collective::{Module, Call, Event<T>, Origin<T>, Config<T>},
+	}
```
Comment (thiolliere) [inline frame/collective/src/mock.rs:151]: If you use the metadata constructed here, then you should add Storage to be able to compare the storage metadata from before to after
```suggestion
		Collective: collective::<Instance1>::{Module, Call, Storage, Event<T>, Origin<T>, Config<T>},
		CollectiveMajority: collective::<Instance2>::{Module, Call, Storage, Event<T>, Origin<T>, Config<T>},
		DefaultCollective: collective::{Module, Call, Storage, Event<T>, Origin<T>, Config<T>},
	}
```

File: frame/collective/src/lib.rs [hunk 5]
```diff
@@ -683,29 +499,197 @@ decl_module! {
...
+
+	/// The current members of the collective. This is stored sorted (just by value).
+	#[pallet::storage]
+	#[pallet::getter(fn members)]
+	pub type Members<T: Config<I>, I: 'static = ()> = StorageValue<_, Vec<T::AccountId>, ValueQuery>;
+
+	/// The prime member that helps determine the default vote behavior in case of absentations.
+	#[pallet::storage]
+	#[pallet::getter(fn prime)]
+	pub type Prime<T: Config<I>, I: 'static = ()> = StorageValue<_, Option<T::AccountId>, ValueQuery>;
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:620]:
```suggestion
	pub type Prime<T: Config<I>, I: 'static = ()> = StorageValue<_, T::AccountId, OptionQuery>;
```
This should appear in the metadata, if the pallet has `Storage` part in construct_runtime

File: frame/collective/src/lib.rs [hunk 6]
```diff
@@ -683,29 +499,197 @@ decl_module! {
...
+
+	/// Votes on a given proposal, if it is ongoing.
+	#[pallet::storage]
+	#[pallet::getter(fn voting)]
+	pub type Voting<T: Config<I>, I: 'static = ()> = StorageMap<
+		_,
+		Identity,
+		T::Hash,
+		Option<Votes<T::AccountId, T::BlockNumber>>,
+		ValueQuery
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:604]:
```suggestion
		Votes<T::AccountId, T::BlockNumber>,
		OptionQuery
```

File: frame/collective/src/lib.rs [hunk 7]
```diff
@@ -683,29 +499,197 @@ decl_module! {
...
+		/// \[proposal_hash, result\]
+		MemberExecuted(T::Hash, DispatchResult),
+		/// A proposal was closed because its threshold was reached or after its duration was up.
+		/// \[proposal_hash, yes, no\]
+		Closed(T::Hash, MemberCount, MemberCount),
+	}
+
+	/// Old name generated by `decl_event`.
+	// TODO :: Have to re-check is "Note message" required ?
+	// #[deprecated(note = "use `Event` instead")]
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:553]: better to make it deprecated IMO

File: frame/collective/src/lib.rs [hunk 8]
```diff
@@ -427,31 +261,35 @@ decl_module! {
...
 			let proposal_len = proposal.using_encoded(|x| x.len());
-			ensure!(proposal_len <= length_bound as usize, Error::<T, I>::WrongProposalLength);
+			ensure!(proposal_len <= length_bound as usize, <Error<T, I>>::WrongProposalLength);
 			let proposal_hash = T::Hashing::hash_of(&proposal);
-			ensure!(!<ProposalOf<T, I>>::contains_key(proposal_hash), Error::<T, I>::DuplicateProposal);
+
+			ensure!(
+				!<ProposalOf<T, I>>::contains_key(proposal_hash),
+				<Error<T, I>>::DuplicateProposal
+			);
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:283]: these refactor should be done in another PR

File: frame/collective/src/lib.rs [hunk 9]
```diff
@@ -91,195 +97,64 @@ pub trait DefaultVote {
...
-	}
-}
+	/// Origin for the collective module.
+	#[pallet::origin]
+	pub type Origin<T, I = ()> = RawOrigin<<T as frame_system::Config>::AccountId, I>;
 
-/// Return the weight of a dispatch call result as an `Option`.
-///
-/// Will return the weight regardless of what the state of the result is.
-fn get_result_weight(result: DispatchResultWithPostInfo) -> Option<Weight> {
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:269]: why did you move this method to inside the Pallet ?
I think better not to do these refactor in this PR

File: frame/collective/src/lib.rs (hunk 2 above)
Comment (thiolliere) [reply frame/collective/src/lib.rs:961]: at some point I would prefer not to do the refactor in this PR, so it is easier to review

Comment (shamb0) [reply frame/collective/src/lib.rs:961]: I assumed even pallet file structure needs to be refactored, sorry for my bad assumption :+1: ,

Shall I move the test back to `lib.rs` ? May be as suggested, will take file restructure part of  another PR

File: frame/collective/src/lib.rs (hunk 8 above)
Comment (thiolliere) [reply frame/collective/src/lib.rs:283]: actually this refactor is fine, sorry for having been too nitpicking

File: frame/collective/src/benchmarking.rs (hunk 3 above)
Comment (thiolliere) [reply frame/collective/src/benchmarking.rs:49]: ok it is not supported, waiting on https://github.com/paritytech/substrate/pull/8190
